src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/deciders/jvm/old_gen/LevelTwoActionBuilder.java

```diff
@@ -0,0 +1,183 @@
+
+  private LevelTwoActionBuilder(final NodeKey esNode, final NodeConfigCache nodeConfigCache) {
+    super(esNode, nodeConfigCache);
+  }
+
+  public static LevelTwoActionBuilder newBuilder(final NodeKey esNode, final NodeConfigCache nodeConfigCache) {
+    return new LevelTwoActionBuilder(esNode, nodeConfigCache);
+  }
+
+  private void addFieldDataCacheAction() {
+    Long capacity = NodeConfigCacheReaderUtil
+        .readCacheMaxSizeInBytes(nodeConfigCache, esNode, ResourceEnum.FIELD_DATA_CACHE);
+    if (capacity == null) {
+      return;
+    }
+    ModifyCacheMaxSizeAction action = new ModifyCacheMaxSizeAction(esNode, ResourceEnum.FIELD_DATA_CACHE,
+        nodeConfigCache, capacity, false, LEVEL_TWO_CONST.CACHE_ACTION_STEP_COUNT);
+    if (action.isActionable()
+        && action.getDesiredCapacityInPercent() >= LEVEL_TWO_CONST.FIELD_DATA_CACHE_LOWER_BOUND) {
+      cacheActionMap.put(ResourceEnum.FIELD_DATA_CACHE, action);
+    }
+  }
+
+  private void addShardRequestCacheAction() {
+    Long capacity = NodeConfigCacheReaderUtil
+        .readCacheMaxSizeInBytes(nodeConfigCache, esNode, ResourceEnum.SHARD_REQUEST_CACHE);
+    if (capacity == null) {
+      return;
+    }
+    ModifyCacheMaxSizeAction action = new ModifyCacheMaxSizeAction(esNode, ResourceEnum.SHARD_REQUEST_CACHE,
+        nodeConfigCache, capacity, false, LEVEL_TWO_CONST.CACHE_ACTION_STEP_COUNT);
+    if (action.isActionable()
+        && action.getDesiredCapacityInPercent() >= LEVEL_TWO_CONST.SHARD_REQUEST_CACHE_LOWER_BOUND) {
+      cacheActionMap.put(ResourceEnum.SHARD_REQUEST_CACHE, action);
+    }
+  }
+
+  private void addWriteQueueAction() {
+    Integer capacity = NodeConfigCacheReaderUtil
+        .readQueueCapacity(nodeConfigCache, esNode, ResourceEnum.WRITE_THREADPOOL);
+    if (capacity == null) {
+      return;
+    }
+    ModifyQueueCapacityAction action = new ModifyQueueCapacityAction(esNode, ResourceEnum.WRITE_THREADPOOL,
+        capacity, false, LEVEL_TWO_CONST.QUEUE_ACTION_STEP_COUNT);
+    if (action.isActionable()) {
+      queueActionMap.put(ResourceEnum.WRITE_THREADPOOL, action);
+    }
+  }
+
+  private void addSearchQueueAction() {
+    Integer capacity = NodeConfigCacheReaderUtil
+        .readQueueCapacity(nodeConfigCache, esNode, ResourceEnum.SEARCH_THREADPOOL);
+    if (capacity == null) {
+      return;
+    }
+    ModifyQueueCapacityAction action = new ModifyQueueCapacityAction(esNode, ResourceEnum.SEARCH_THREADPOOL,
+        capacity, false, LEVEL_TWO_CONST.QUEUE_ACTION_STEP_COUNT);
+    if (action.isActionable()) {
+      queueActionMap.put(ResourceEnum.SEARCH_THREADPOOL, action);
+    }
+  }
+
+
+  private void actionPriorityForCache() {
+    actionFilter.put(ResourceEnum.FIELD_DATA_CACHE, true);
+    actionFilter.put(ResourceEnum.SHARD_REQUEST_CACHE, true);
+  }
+
+  // allocate value to its bucket
+  private int bucketization(int lowerBound, int upperBound, int val, int bucketSize) {
```

Comment:
Add a doc string to explain this formulae, and what `step` means here

Reply:
done